src/adapter/src/catalog/migrate.rs

```diff
@@ -182,6 +210,576 @@ pub(crate) async fn migrate(
+                let ids = tx.allocate_user_item_ids(1)?;
+                let (new_table_id, new_table_global_id) = ids[0];
+                info!("migrate: added table {new_table_id}: {table}");
+                tx.insert_user_item(
+                    new_table_id,
+                    new_table_global_id,
+                    schema_id,
+                    &table_item_name,
+                    table.to_ast_string_stable(),
+                    owner_id,
+                    privileges,
+                    &Default::default(),
+                    extra_versions,
+                )?;
+                // We need to move the shard currently attached to the source statement to the
+                // table statement such that the existing data in the shard is preserved and can
+                // be queried on the new table statement. However, we need to keep the GlobalId of
+                // the source the same, to preserve existing references to that statement in
+                // external tools such as DBT and Terraform. We will insert a new shard for the source
+                // statement which will be automatically created after the migration is complete.
+                let new_source_shard = ShardId::new();
+                let (source_global_id, existing_source_shard) = tx
+                    .delete_collection_metadata(btreeset! {source_global_id})
+                    .pop()
+                    .expect("shard should exist");
+                tx.insert_collection_metadata(btreemap! {
+                    new_table_global_id => existing_source_shard,
+                    source_global_id => new_source_shard
+                })?;
+
+                add_to_audit_log(
+                    tx,
+                    mz_audit_log::EventType::Create,
+                    mz_audit_log::ObjectType::Table,
+                    mz_audit_log::EventDetails::IdFullNameV1(mz_audit_log::IdFullNameV1 {
+                        id: new_table_id.to_string(),
+                        name: mz_audit_log::FullNameV1 {
+                            database: schema
+                                .database_id
+                                .map(|d| d.to_string())
+                                .unwrap_or_default(),
+                            schema: schema.name,
+                            item: table_item_name,
+                        },
+                    }),
+                    now(),
+                )?;
+
+                // We also need to update any other statements that reference the source to use the new
+                // table id/name instead.
+                changed_ids.insert(source_id, new_table_id);
+            }
+
+            // When we upgrade to > rust 1.81 we should use #[expect(unreachable_patterns)]
+            // to enforce that we have covered all CreateSourceStatement variants.
+            #[allow(unreachable_patterns)]
+            Statement::CreateSource(_) => {}
```

Comment:
We now use Rust 1.83 so you could add an `#[expect(unreachable_patterns)]` here if you wanted

Reply:
Done